### PR TITLE
Move block ID remaps from external PHP script into core Java code.

### PR DESCRIPTION
**Pull request by magicus**
I have verified programmatically that the resulting remaps are identical to what was produced by the old code. (Verification code removed before final commit.)

### CODE REVIEW

**Comment by magicus** [comment]
@Wesley-VanNeck This is a rewrite to have all remapping logic in the Java code base, instead of having it in a separate pop script.

File: src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java [hunk 1]
```diff
@@ -137,9 +140,177 @@ public static NetworkEntityType getLivingTypeFromPeNetworkId(int networkId) {
 	}
 
 	public static final ArrayBasedIdRemappingTable BLOCK_ID = new ArrayBasedIdRemappingTable(MinecraftData.BLOCK_ID_MAX * MinecraftData.BLOCK_DATA_MAX);
+
+	private static HashMap<Integer, ArrayList<Integer>> blockIdRemaps = new HashMap<>();
```
Comment (yannicklamprecht) [inline src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java]: Follow LSP.

File: src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java [hunk 2]
```diff
@@ -137,9 +140,177 @@ public static NetworkEntityType getLivingTypeFromPeNetworkId(int networkId) {
...
+	private static HashMap<Integer, ArrayList<Integer>> blockStateRemaps = new HashMap<>();
```
Comment (yannicklamprecht) [inline src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java]: LSP

File: src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java [hunk 3]
```diff
@@ -137,9 +140,177 @@ public static NetworkEntityType getLivingTypeFromPeNetworkId(int networkId) {
...
+		blocksJson.forEach(entry -> {
+			// For most blocks, map PC blockState to PE runtime ID. For some blocks,
+			// the PE id/blockState is different from the PC id/blockState. These
+			// are encoded in our remaps.
+			int runtimeID = entry.getAsJsonObject().get("runtimeID").getAsInt();
+			int id = entry.getAsJsonObject().get("id").getAsInt(); // PE id
+			int data = entry.getAsJsonObject().get("data").getAsInt(); // PE data
+
+			int blockState = (id << 4) | data;
+			ArrayList<Integer> remappedIds = blockIdRemaps.get(id);
```
Comment (yannicklamprecht) [inline src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java]: LSP

File: src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java [hunk 4]
```diff
@@ -137,9 +140,177 @@ public static NetworkEntityType getLivingTypeFromPeNetworkId(int networkId) {
...
+			ArrayList<Integer> remappedBlockStates = blockStateRemaps.get(blockState);
```
Comment (yannicklamprecht) [inline src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java]: LSP

**Comment by 7kasper** [comment]
Thank you. I will be porting it myself as to include the new changes for 1.13.

**Comment by yannicklamprecht** [comment]
By the way it’s JSON configuration file. I see only advantages using the data driven way.

File: src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java (hunk 1 above)
Comment (Shevchik) [reply src/protocolsupport/protocol/typeremapper/pe/PEDataValues.java]: We don't require following LSP actually, but this field is actually missing an final modifier, which is more important IMO.

**Comment by yannicklamprecht** [comment]
@7kasper with LSP or with data driven approach?

LSP -> Liskov Substitution Principle -> develop against abstraction instead of implementation. 

data driven approach -> instead of hard coding everything you can add configuration to improve adaptation speed.

**Pull request closed without merging**